packages/actions-shared/src/engage/utils/EngageActionPerformer.ts

```diff
@@ -168,6 +184,344 @@ export abstract class EngageActionPerformer<TSettings = any, TPayload = any, TRe
+            }
+          }
+        )
+      )
+
+      if (cacheSavingError || !cacheSavingResult) {
+        options?.onSaveFailed?.(cacheSavingError)
+      }
+    } else {
+      step.write('cache_save_skipped')
+    }
+
+    if (resultError) throw resultError
+    else return result as T
+  }
+
+  /**
+   * Distributed lock implementation using Redis
+   *
+   * LOGIC:
+   * Trying to aquire lock. Possible outcomes:
+   * 1. Lock acquired:
+   *    - createValue() and finally release lock
+   * 2. Lock not acquired because of timeout:
+   *   - throw timeout error up
+   * 3. Lock not acquired because of error accessing redis:
+   *   - throw retryable error //fallback to createValue()
+   * @param key resource key to lock under
+   * @param createValue function to execute if lock is acquired
+   * @param options lock options
+   * @returns
+   */
+  @track()
+  async withDistributedLock<T>(key: string, createValue: () => Promise<T>, options: LockOptions): Promise<T> {
+    const cache_group = options.cacheGroup || this.currentOperation?.parent?.func.name || ''
+    const step = this.createStepLogger({
+      logs: {
+        key
+      },
+      tags: {
+        cache_group
+      }
+    })
+
+    if (!this.engageDestinationCache) {
+      return await createValue()
+    }
+    const cache = this.engageDestinationCache
+
+    const lockKey = `lock:${key}`
+    const acquireLock = async () => {
+      //tries to acquire lock for acquireLockMaxWaitInSeconds seconds
+      const startTime = Date.now()
+      while (Date.now() - startTime < options.acquireLockMaxWaitTimeMs) {
+        step.write('cache_lock_attempt')
+        if (await step.track('redis_setNX', () => cache.setByKeyNX(lockKey, 'locked', options.lockMaxTimeMs / 1000))) {
+          // lock acquired, returning release function
+          step.write('cache_lock_acquired', { tags: { lock_acquired: true } })
+          this.statsHistogram('cache_lock_waiting_time', Date.now() - startTime, step.tags)
+          return {
+            release: () => step.track('redis_delByKey', () => cache.delByKey(lockKey))
+          }
+        }
+        step.write('cache_lock_waiting')
+        await new Promise((resolve) => setTimeout(resolve, options.acquireLockRetryIntervalMs || 500)) // Wait 500ms before retrying
+      }
+      //no lock acquired because of waiting timeout
+      step.write('cache_lock_timeout')
+    }
+
+    const { value: lock, error: redisError } = await step.track('cache_acquire_lock', () =>
+      getOrCatch(() => acquireLock())
+    )
+
+    if (redisError) {
+      this.throwRetryableError('Error acquiring lock: ' + redisError.message)
+    } else if (!lock?.release) {
+      // no redis error and no lock acquired - means it was acquiring timeout
+      step.write('cache_lock_timeout_exit')
+      this.throwRetryableError('Timeout while acquiring lock')
+    }
+    //if lock obtained or there was redis error - execute createValue and finally release lock
```

Comment:
Seems like you handle the redis error case in the first if block?
```suggestion
    //if lock obtained  - execute createValue and finally release lock
```

Reply:
good catch, there were multuple iterations of changes of this logic, and I forgot to update the comment